### ISSUE TITLE
Azure Support

### ISSUE THREAD

**Issue by chrislovecnm** [kubernetes/kops#3957]
# What - Support Azure

## Problem Statement

Support the running a kops cluster on Azure.  This implementation would model AWS and GCE, and would be awesome!

## How

Azure includes full API support for the creation of virtual machines and their scaling groups.  There are multiple examples of the 'hard way' but we first can model there managed version.

## Why

_Why not_ ?  Having AWS, GCE, Azure is the trifecta

## Design

TBD

**Comment by robinpercy** [comment]
+1 Cloud-agnosticism is a consistent priority I hear from customers deploying K8s.  Being able to use Kops across all three would solve a lot of pain.

**Comment by chrislovecnm** [comment]
@robinpercy I have gotten asks from my clients that they want to run on the big three. We got two down, one to go

**Comment by vendrov** [comment]
Great idea! I believe that our first step should be, adding windows nodes support for kops :)

**Comment by chrislovecnm** [comment]
Yes we do need windows node support as well. Will open a separate ticket. I have talked with Microsoft shops who would love that,

**Comment by vendrov** [comment]
No need to open a new one, https://github.com/kubernetes/kops/issues/2422

**Comment by bgeesaman** [comment]
I've looked at over a dozen K8s installation tools and managed K8s services in the past few months on all major providers.  What was painfully clear was the level of fragmentation in the tools and offerings as well as what features are supported.  It required a ton of mental energy to translate terms and patterns to get similar output.

We are at a point where focused efforts like this can provide that next level of true abstraction so that administrators/operators can see and manage Kubernetes uniformly--regardless of cloud provider.  I think bringing all the advanced features and capabilities (HA, rolling-upgrades, etc) battle-tested in AWS and GCE to Azure is a no-brainer.

**Comment by chrislovecnm** [comment]
@bgeesaman well put - fortunately the coming machine api should help with this, but the installers do have different use cases

**Comment by geojaz** [comment]
We should be building out azure support regardless of windows nodes. We can definitely consider building out windows node support in kops, but I would start with the basics. That is to say, let's build out linux support first because that's what we support on AWS and GCE. That will bring some level of feature parity to the 3 biggies via kops.

I too have clients that need 1 tool to manage all the clusters across these providers and kops is a great way to go. We now have much of the background work done to build a new provider so I think it's quite doable. I'm personally not that up to speed on Azure, but I'd like to see a good "azure-native" implementation to complement aws and gcp. Excited to see this get attention and I'm happy to help.

**Comment by robertojrojas** [comment]
IMHO, Azure support is a must have! I agree with @geojaz windows support would be nice, but Linux should come first. I'll be interesting to see what changes would be needed in the CLI and API to accomplish this. I'm happy to help as well.

**Comment by brandoncole** [comment]
This would be great.  There are definitely some companies who prefer Azure to AWS in foreign countries which is where this comes into play a lot.  In the U.S.A. there's indifference to some degree in the cloud vendor, but there's certain places globally where the choice has been made for you which we have to respect.

**Comment by justinsb** [comment]
Yes, I'd love to see us work on this.  Azure is doing some very cool stuff, and it would be great to help people to try out Kubernetes on Azure as easily as they can try out AWS & now GCE.

**Comment by hubt** [comment]
This would be excellent. Windows support isn't important at least initially. People need this for cloud portability(usually because of cost), not OS portability.

**Comment by perryao** [comment]
@chrislovecnm Is this work active or still awaiting the machine api?

**Comment by chrislovecnm** [comment]
@perryao nobody has started on this, would love someone to pick this up!

**Comment by perryao** [comment]
I'd like to try my hand at this. I'll work on a rough version over the coming weeks and hopefully come back to discuss design with you @chrislovecnm

**Comment by chrislovecnm** [comment]
@perryao let me know, if you have questions happy to chat with you.

**Comment by rajeevb2** [comment]
@perryao any news on kops for azure?

**Comment by perryao** [comment]
@rajeevb2 unfortunately not too much. I have started and have a pretty good feel for what needs to be done, but haven't had enough time to make meaningful progress on it yet. But still a priority for me.

At this point I've just stubbed out the tasks I'll need in cloudup and starting implementing a few interfaces.

**Comment by chrislovecnm** [comment]
@perryao if you want to add a feature flag, please start pushing code.  Love to have more than one person working on this if you feel it would help.

Sometimes more people don't help 😆

**Comment by mliyanagamage** [comment]
Happy to help with testing if you need it

**Comment by marc-sensenich** [comment]
@perryao If you would like to have some help on this from the testing and development front, I'd be willing to help out

**Comment by perryao** [comment]
@chrislovecnm @marc-sensenich thanks guys. I just pushed the initial stuff I did to layout the changes I'd need to make. Here's the diff of where I'm at now: https://github.com/kubernetes/kops/compare/master...perryao:feature/%233957azure-support?expand=1. I'm wrapping up a deliverable at work this week, but I think I should be back to working on this early next week.

**Comment by chezbut** [comment]
Is there any plan for being able to run this on local Hyper-V hosts, with multiple VM's?

**Comment by Cittarasu** [comment]
@chrislovecnm @perryao  -- great start on Azure front. I can help in testing/development front. Let me know how we can sync-up on this front. We have been using KOPS for AWS extensively but we are exploring KOPS for vsphere and Azure.

**Comment by luisdavim** [comment]
Any updates on this?

**Comment by thangchung** [comment]
+1

**Comment by wandergeek** [comment]
+10

**Comment by rajivchodisetti** [comment]
+10

**Comment by billpratt** [comment]
Hello all! I started working on this. So far I have most of the state store logic complete to save to Azure storage. You can follow along here: https://github.com/kubernetes/kops/compare/master...billpratt:azure?expand=1

Cheers!
Bill

**Comment by billpratt** [comment]
Hello everyone. I have started to work on this during free time. You can track progress here: https://github.com/billpratt/kops/commits/azure

So far I have basic state-store functionality working with Azure blob storage.

**Comment by dazdaz** [comment]
Awesome to see the beginning of the work to implement Azure support with kops.
May I ask what the current status is.
Have you considered creating a roadmap of features/comaptibility to highlight where we're at and what requires developer effort.

**Comment by billpratt** [comment]
@dazdaz I havent had as much time lately to work on this. I plan to start up again soon. I originally created [this project](https://github.com/billpratt/kops/projects/2) on my fork. It needs a little bit of cleanup. I also ran into a blocker where a downstream dependency of kops requires a specific version of the azure-sdk-for-go. I will need to update to use a newer version. I need to also evaluate what version of k8s to start supporting this in Azure. Kubernetes 1.12 released support for Azure VM Scale Sets, which gives similar functionality as AWS scale sets. If kops is going to support < 1.12 for Azure, I'll have to use Azure Availability Sets which does not give you automatic cluster scaling.

**Comment by argiris-mat** [comment]
@billpratt hello there. Can we do anything to support this? Is there anything I could test or document?

**Comment by billpratt** [comment]
Not yet. I have some free time coming up and I plan to continue to work on this. I will let you know when something is ready.

**Comment by igormusinov** [comment]
any updates?

**Comment by innlouvate** [comment]
Any chance this might be coming in the next 6 months or so? Would happily support if I can.

**Comment by bzuelke** [comment]
i need this in my life so bad. I hope in the near future this can become a reality. I hate having to force myself to have a snowflake in Azure from our AWS clusters.

**Comment by DG4227** [comment]
Would love to see this come to fruition!

**Comment by bzuelke** [comment]
@billpratt I've looked over your fork and looks like you've done a ton of leg work. Is there anything that anyone here could do to try and get this in a PR? Didn't want to reinvent the wheel if I didn't have to.

**Comment by bryaneburr** [comment]
MS loves to give away free time on azure, and higher-ups love to take it. This would make devops' lives like a billion times easier 😀

**Comment by billpratt** [comment]
Hi everyone,

This was strictly a personal project of mine that I worked on at night and **no** work was done as a Microsoft employee. Unfortunately I have had no time finish. @bzuelke, if you or anyone else on this thread would like to fork what I've done and take over, please do as I would also love to see this feature added 😃

**Comment by bzuelke** [comment]
Thanks for getting back to the thread @billpratt I totally get where you're at. I'm kind of in the same boat time wise which is frustrating naturally. I know you have a decent amount of leg work put into the code and if the universe is willing to grant me some time to check it out I'll 100% fork it and muck with it. Or if someone else jumps on it I'm totally open to collab to work this out.

**Comment by pierluigilenoci** [comment]
@billpratt any news?

**Comment by bzuelke** [comment]
Chances are there won't be any support for Azure unless someone really jumps on it. But I think more people use other forms of cluster management until Cluster API goes GA or simply uses AKS. I would have like to take a stab at this but simply don't have resources available to follow up on Bills work :/

**Comment by christianh814** [comment]
> Chances are there won't be any support for Azure unless someone really jumps on it. But I think more people use other forms of cluster management until Cluster API goes GA or simply uses AKS. I would have like to take a stab at this but simply don't have resources available to follow up on Bills work :/

+1 to this comment and the ":/" ...realistically; when Cluster API matures, KOPS would/should use that as the core of the tool.

**Comment by rendhalver** [comment]
Just catching up on this. I see cluster-api has a non pre-release release in [0.28.0](https://github.com/kubernetes-sigs/cluster-api/releases/tag/v0.2.8).
Is that what people were waiting for?
I am also interested in helping out where I can.
I have been looking for reason to learn more go and we use kops to deploy our clusters and having the ability to manage all of them with kops would be grand.

**Comment by romantsai-github** [comment]
any updated? like to know how kops support on Azure if possible?

**Comment by Cittarasu** [comment]
Wondering why KOPS support require - as each cloud provider now providing
k8s engine via API or CLI support.

**Comment by lsca** [comment]
Any updates on this?

**Comment by lsca** [comment]
Any updates on this?

**Comment by bzuelke** [comment]
Mentioned before this will probably not be a thing unless someone jumps on it. I don't believe anyone is currently working on this or have initiative. IIRC cluster API will be a defacto for Azure. I know Kops will begin utilizing CPAI based on the talk they gave at kubecon 2019 but still isn't a thing yet and will be in phases. Also at said talk they mentioned efforts on other aspects of kops so thinking main contributors aren't focusing efforts on Azure and might never. Someone will most likely need to hack through this or use another tooling like AKS-Engine or CAPI. I mention this considering the previous conversations in this thread and the age of it.

**Comment by luisdavim** [comment]
Yeah, focusing on cluster API seems like a better idea that would eventually lead to solving this issue as well.

**Comment by kenji-cloudnatix** [comment]
Just wanted to update this issue as we have a fork that supports Azure.

It's not fully done, but major features have been implemented, including:

- management of Azure resources such as Resource Group, Virtual Network, VM Scale Sets and Managed Disks.
- etcd cluster setup with Azure Managed Disks in etcd-manager
- Azure Cloud config injection
- VFS backed by Azure Blob storage service

We can keep the changes in the fork, but we would like to upstream it if possible. Is there any interest in the community? Here are the change diffs:

- kops: https://github.com/cloudnatix/kops-public/compare/base...cloudnatix:azure
- etcd-manager: https://github.com/cloudnatix/etcd-manager-public/compare/base...cloudnatix:azure

**Comment by bzuelke** [comment]
personally id love to see support for this even if there is a shift toward cluster-api. this is amazing and can be added onto if needed. haven't checked the code base yet but this would be great to have 1 tool for multi cloud provider if we can.

**Comment by rifelpet** [comment]
Great work @kenji-cloudnatix ! Would you or someone from your team be able to join the next [Kops office hours](https://kops.sigs.k8s.io/welcome/office_hours/)? Friday Oct 23rd at 12pm US/Eastern. We can discuss a high level plan with other maintainers there and then hash out any details here or in PRs.

**Comment by hakman** [comment]
Would be really cool to get this into Kops. Thanks @kenji-cloudnatix !

**Comment by kenji-cloudnatix** [comment]
Thanks all! Sure, will join the office hours on Oct 23rd.

**Comment by rifelpet** [comment]
@kenji as discussed today, I think our next steps will be:
1. Open the big kops PR with all the changes. we'll likely end up breaking it up into smaller PRs but it helps to see the full scope of changes.
2. Open a kops PR with just the VFS changes (`util/pkg/vfs/*`) + necessary dependencies
3. Once the VFS PR is merged, open the etcd-manager PR with the updated kops dependency

In the big PR we can further discuss how to break down changes as necessary.

**Comment by rifelpet** [comment]
@kenji-cloudnatix as discussed today, I think our next steps will be:
1. Open the big kops PR with all the changes. we'll likely end up breaking it up into smaller PRs but it helps to see the full scope of changes.
2. Open a kops PR with just the VFS changes (`util/pkg/vfs/*`) + necessary dependencies
3. Once the VFS PR is merged, open the etcd-manager PR with the updated kops dependency

In the big PR we can further discuss how to break down changes as necessary.

**Comment by kenji-cloudnatix** [comment]
Sure, @rifelpet ! Just opened https://github.com/kubernetes/kops/pull/10114 for the first item.

**Comment by hakman** [comment]
Azure support is close to graduating to Beta.
/close
